### PR TITLE
Identify datasets in nested structures

### PR DESCRIPTION
**Pull request by sven1103**
This PR introduces a DatasetLocatorImpl class, that enables the ETL
to grab the actual dataset path in case of nested structures, generated after processing with the Dropboxhandler.

### CODE REVIEW

**Comment by KochTobi** [comment]
Is this ready for review?

File: src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    @Shared
+    String flatDataStructureExample
+
+    def setup() {
+        setupNestedDataset()
+        setupRelativePathExample()
+        setupFlatDatasetExample()
+    }
+
+    def "Non absolute path shall return an IllegalArgumentException"() {
```
Comment (KochTobi) [inline src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy:30]:
```suggestion
    def "Given a relative path, when a DatasetLocator is created then an IllegalArgumentException is thrown"() {
```

File: src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        when:
+        DatasetLocator locator = DatasetLocatorImpl.of(relativePathExample)
+
+        then:
+        thrown(IllegalArgumentException.class)
+    }
+
+    def "Given an absolute path shall successfully create an instance of this class"() {
```
Comment (KochTobi) [inline src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy:38]:
```suggestion
    def "Given an absolute path, when a DatasetLocator is created then an instance of this class is returned"() {
```

File: src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        when:
+        DatasetLocator locator = DatasetLocatorImpl.of(nestedDataStructureExample)
+
+        then:
+        noExceptionThrown()
+    }
+
+    def "Given an nested structure, return the actual dataset absolute path"() {
```
Comment (KochTobi) [inline src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy:46]:
```suggestion
    def "Given a nested structure, when the path to the dataset folder is requested, return the absolute path to the parent folder"() {
```

File: src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        when:
+        String datasetPath = locator.getPathToDatasetFolder()
+
+        then:
+        datasetPath != nestedDataStructureExample
+
+        new File(datasetPath).getParent().equalsIgnoreCase(nestedDataStructureExample)
+    }
+
+    def "Given an flat structure, return the actual root dataset absolute path"() {
```
Comment (KochTobi) [inline src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy:59]:
```suggestion
    def "Given a flat structure, when the path to the dataset folder is requested, return the given path"() {
```

**Review by KochTobi — COMMENTED** [review]
I think it would be nice to change the test names :+1: Very nice to see the tests here and directly understand what the behaviour should be!

File: src/main/groovy/life/qbic/registration/MainETL.groovy [hunk 5]
```diff
@@ -3,36 +3,36 @@ package life.qbic.registration
...
+import java.nio.file.Paths
 
 @Log4j2
 class MainETL extends AbstractJavaDataSetRegistrationDropboxV2 {
 
     static List<DatasetParser<?>> listOfParsers = [
-            new BioinformaticAnalysisParser(),
-            new MaxQuantParser()
+            new MaxQuantParser(),
+            new BioinformaticAnalysisParser()
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/registration/MainETL.groovy:20]: Why the change in sequence? Is it important in which order they are checked?

File: src/main/groovy/life/qbic/registration/MainETL.groovy [hunk 6]
```diff
@@ -46,17 +46,19 @@ class MainETL extends AbstractJavaDataSetRegistrationDropboxV2 {
...
     }
 
-    private static logExceptionReport(List<Exception> observedExceptions) {
+    private static logExceptionReport(Map<String, Exception> observedExceptions) {
         log.error("Detailed exception report:")
         log.error("start---------------------")
-        for (Exception exception : observedExceptions) {
+        for (String parser : observedExceptions.keySet()) {
+            log.error("Report for parser: " + parser)
+            Exception exception = observedExceptions.get(parser)
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/registration/MainETL.groovy:61]: I like the idea with the map :+1:
Just as a note: An alternative way to write this is
```suggestion
        for (def entry : observedExceptions.entrySet()) {
            String parser = entry.key
            Exception exception = entry.value
            log.error("Report for parser: " + parser)
```

File: src/main/groovy/life/qbic/registration/handler/DatasetLocator.groovy [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ *  becomes:
+ *
+ *  |- mydataset_folder
+ *      |- source_dropbox.txt
+ *      |- mydataset_folder
+ *            |- data1.txt
+ *            |- data2.txt
+ *
+ * @since
+ */
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/registration/handler/DatasetLocator.groovy:27]: Very nice JavaDoc ❤️ 
However, I am missing the value for `@since`

File: src/main/groovy/life/qbic/registration/handler/DatasetLocator.groovy [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+interface DatasetLocator {
+
+    /**
+     * Returns the top level absolute path of the dataset. In a flat structure, it returns
+     * the top level folder, in the assumption that the dataset folder is not nested.
+     * @return the absolute dataset folder path
+     */
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/registration/handler/DatasetLocator.groovy:34]: Nit: it would be nice to have a since annotation here as well.

File: src/main/groovy/life/qbic/registration/handler/DatasetLocator.groovy [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ *  becomes:
+ *
+ *  |- mydataset_folder
+ *      |- source_dropbox.txt
+ *      |- mydataset_folder
+ *            |- data1.txt
+ *            |- data2.txt
+ *
+ * @since
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/registration/handler/DatasetLocator.groovy:26]:
```suggestion
 * @since 1.3.0
```

File: src/main/groovy/life/qbic/registration/handler/DatasetLocator.groovy [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ */
+interface DatasetLocator {
+
+    /**
+     * Returns the top level absolute path of the dataset. In a flat structure, it returns
+     * the top level folder, in the assumption that the dataset folder is not nested.
+     * @return the absolute dataset folder path
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/registration/handler/DatasetLocator.groovy:33]:
```suggestion
     * @return the absolute dataset folder path
     * @since 1.3.0
```

File: src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+     * @inheritDocs
+     */
+    @Override
+    String getPathToDatasetFolder() {
+        return findNestedDataset().orElse(this.rootPath)
+    }
+
+    private Optional<String> findNestedDataset() {
+        File rootDir = new File(rootPath)
+        String dirName = rootDir.getName()
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy:41]: Would it make sense to check whether we really have a directory here?
```suggestion
        if (!rootDir.isDirectory()) throw new RuntimeException("The root file is no directory!")
        String dirName = rootDir.getName()
```

File: src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            if (child.equalsIgnoreCase(dirName)) {
+                return Optional.of(rootDir.getAbsolutePath() + "/" + child)
+            }
+        }
+        return Optional.empty()
+    }
+
+    private static boolean isAbsolutePath(String path) {
+        return new File(path).isAbsolute()
+    }
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy:57]: Could be changed to validate that we get a directory indeed.
```suggestion
    private static boolean isValidRoot(String path) {
        File file = new File(path)
        return file.isAbsolute() && file.isDirectory()
    }
```

File: src/main/groovy/life/qbic/registration/handler/DatasetParserHandler.groovy [hunk 13]
```diff
@@ -63,13 +63,13 @@ class DatasetParserHandler {
...
+            DatasetParser parser = dataSetIterator.next()
             try {
-                DatasetParser parser = dataSetIterator.next()
                 result = Optional.of(parser.parseFrom(root))
             } catch (DataParserException e) {
-                observedExceptions << e
+                observedExceptions.put(parser.getClass().getCanonicalName(), e)
             } catch (DatasetValidationException e){
-                observedExceptions << e
+                observedExceptions.put(parser.getClass().getCanonicalName(), e)
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/registration/handler/DatasetParserHandler.groovy:72]: Very nice, readable and intuitive solution :+1:

File: src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+import spock.lang.Shared
+import spock.lang.Specification
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (KochTobi) [inline src/test/groovy/life/qbic/registration/handler/DatasetLocatorImplSpec.groovy:12]:
```suggestion
```
I don't think a specification needs this as it is pretty self-explanatory in this case.

**Review by KochTobi — CHANGES_REQUESTED** [review]
Hey @sven1103 
Really a nice addition!
Maybe you could address one question I have about the reordering of the parsers and decide on whether we should check for directories or not. I expect a NullPointerException otherwise when you iterate with a for loop over `null`

File: src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy (hunk 11 above)
Comment (sven1103) [reply src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy:41]: Strictly not an requirement, it can be a single file

File: src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy (hunk 12 above)
Comment (sven1103) [reply src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy:57]: SRP, this is not the task of this class :) It just needs to know, that the passed path is absolute

File: src/main/groovy/life/qbic/registration/handler/DatasetParserHandler.groovy (hunk 13 above)
Comment (sven1103) [reply src/main/groovy/life/qbic/registration/handler/DatasetParserHandler.groovy:72]: thanks a lot!

File: src/main/groovy/life/qbic/registration/MainETL.groovy (hunk 6 above)
Comment (sven1103) [reply src/main/groovy/life/qbic/registration/MainETL.groovy:61]: I think there are many possible permutations to do this, that was just my flavor of the day :D

File: src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy (hunk 11 above)
Comment (KochTobi) [reply src/main/groovy/life/qbic/registration/handler/DatasetLocatorImpl.groovy:41]: In case this is no directory the `rootDir.list()` call resolves to `null`.
Here I would expect a NullPointerException. Thus I think it is a requirement for this method.

**Review by KochTobi — APPROVED** [review]
LGTM
